Utilities/rANS/include/librans/Coder.h

```diff
@@ -0,0 +1,269 @@
+#include <cassert>
+
+#include "DecoderSymbol.h"
+#include "EncoderSymbol.h"
+#include "helper.h"
+
+namespace o2
+{
+namespace rans
+{
+
+// State for a rANS encoder. Yep, that's all there is to it.
+template <typename T>
+using State = T;
+
+__extension__ typedef unsigned __int128 uint128;
+
+// READ ME FIRST:
+//
+// This is designed like a typical arithmetic coder API, but there's three
+// twists you absolutely should be aware of before you start hacking:
+//
+// 1. You need to encode data in *reverse* - last symbol first. rANS works
+//    like a stack: last in, first out.
+// 2. Likewise, the encoder outputs bytes *in reverse* - that is, you give
+//    it a pointer to the *end* of your buffer (exclusive), and it will
+//    slowly move towards the beginning as more bytes are emitted.
+// 3. Unlike basically any other entropy coder implementation you might
+//    have used, you can interleave data from multiple independent rANS
+//    encoders into the same bytestream without any extra signaling;
+//    you can also just write some bytes by yourself in the middle if
+//    you want to. This is in addition to the usual arithmetic encoder
+//    property of being able to switch models on the fly. Writing raw
+//    bytes can be useful when you have some data that you know is
+//    incompressible, and is cheaper than going through the rANS encode
+//    function. Using multiple rANS coders on the same byte stream wastes
+//    a few bytes compared to using just one, but execution of two
+//    independent encoders can happen in parallel on superscalar and
+//    Out-of-Order CPUs, so this can be *much* faster in tight decoding
+//    loops.
+//
+//    This is why all the rANS functions take the write pointer as an
+//    argument instead of just storing it in some context struct.
+
+// --------------------------------------------------------------------------
+template <typename T, typename Stream_t>
+class Coder
+{
+ public:
+  Coder() = delete;
+
+  // Initialize a rANS encoder.
+  static void encInit(State<T>* r)
+  {
+    *r = LOWER_BOUND;
+  };
+
+  // Encodes a single symbol with range start "start" and frequency "freq".
+  // All frequencies are assumed to sum to "1 << scale_bits", and the
+  // resulting bytes get written to ptr (which is updated).
+  //
+  // NOTE: With rANS, you need to encode symbols in *reverse order*, i.e. from
+  // beginning to end! Likewise, the output bytestream is written *backwards*:
+  // ptr starts pointing at the end of the output buffer and keeps decrementing.
+  static void encPut(State<T>* r, Stream_t** pptr, uint32_t start, uint32_t freq, uint32_t scale_bits)
+  {
+    // renormalize
+    State<T> x = encRenorm(*r, pptr, freq, scale_bits);
+
+    // x = C(s,x)
+    *r = ((x / freq) << scale_bits) + (x % freq) + start;
+  };
+
+  // Flushes the rANS encoder.
+  static void encFlush(State<T>* r, Stream_t** pptr)
+  {
+    T x = *r;
+    Stream_t* ptr = *pptr;
+
+    if constexpr (needs64Bit<T>()) {
```

Comment:
I don't fully understand the logic here, this decision is purely on the state variable, but doesn't this require a consistency check for `Stream_t` as well? In the true branch it needs to be 4 byte wide, in the false branch 1 byte.

Reply:
Thanks, indeed this makes no sense. There have originally been two different files one for 32 and one for 64 bit states, where they differed it seems that I blindly introduced the ` if constexpr (needs64Bit<T>())` to differenciate between the implementations. I will review them.

Reply:
So indeed the code is not universal but only works with the combination
64 bit arithmetic, 32 bit streaming, 32 bit arithmetic, 8 bit streaming. I really forgot about it because it has been a longer time since I last touched this.
To enforce correct types during compile time I now added a static assert into the `Coder` class. Obviously there are other possibilities and one can think of generalizing the `Coder` further, but it will require some more development and testing to ensure all corner cases are caught. 
For what we do 64Bit arithmeric and 32 bit streaming seems to be the most performant option anyhow.